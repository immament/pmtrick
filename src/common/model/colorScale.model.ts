export class ColorScale {
    private delta: number;
    private lightnessDelta: number;
    constructor(
        public readonly maxValue: number,
        public readonly minValue = 0,
        public readonly hue = 255,
        public readonly saturation = 100,
        public readonly maxLightness = 90,
        public readonly minLightness = 25,
    ) {
        this.delta = this.maxValue - this.minValue;
        this.lightnessDelta = this.maxLightness - this.minLightness;
    }

    // lightness between 25% and 90%
    getHslColor(value: number): string {
        const normalizedValue = Math.min(Math.max(0, value - this.minValue), this.maxValue);
        return `hsl(${this.hue}, ${this.saturation}%, ${
            this.maxLightness - (normalizedValue / this.delta) * this.lightnessDelta
        }%`;
    }
}
