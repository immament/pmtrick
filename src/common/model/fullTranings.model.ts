export class FullTrainings {
    constructor({
        currentSeasonLeagueMatches,
        currentSeasonFriendlyMatches,
        fullSeasonLeagueMatches = 18,
        fullSeasonFriendlyMatches = 12,
    }: {
        currentSeasonLeagueMatches?: number;
        currentSeasonFriendlyMatches?: number;
        fullSeasonLeagueMatches?: number;
        fullSeasonFriendlyMatches?: number;
    }) {
        this.curentSeason = (currentSeasonLeagueMatches ?? 0) + (currentSeasonFriendlyMatches ?? 0) / 2;
        this.fullSeason = fullSeasonLeagueMatches + fullSeasonFriendlyMatches / 2;
    }

    curentSeason: number;
    fullSeason: number;
}
