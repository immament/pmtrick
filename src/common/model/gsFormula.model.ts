export interface GsFormula {
    main: number;
    secondary: number;
    tertiary: number;
    physical: number;
}

export const mainGsFormula: GsFormula = {
    main: 61.02,
    secondary: 20.34,
    physical: 15.25,
    tertiary: 1.13,
};

// TODO
export const tempGsFormula: GsFormula = {
    main: 60,
    secondary: 20,
    physical: 20,
    tertiary: 0,
};
