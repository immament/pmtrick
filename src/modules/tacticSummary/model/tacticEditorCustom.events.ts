export type TacticEditorCustomEvents = 'PMT_tacticInit' | 'PMT_tacticSquadChanged' | 'PMT_advTacticChanged';
