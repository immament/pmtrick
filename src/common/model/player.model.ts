import { MatchesInSeasons } from '@src/common/services/storage/skillsTableOptions.repository';

import { SkillsSummary } from '../../modules/tacticSummary/model/skillsSummary.model';

import { PlayerSkills } from './playerSkills.model';

export type PlayerPositionType = 'G' | 'D' | 'M' | 'F';
export const playerPositionValues: PlayerPositionType[] = ['G', 'D', 'M', 'F'];

export enum PlayerPositionEnum {
    G,
    D,
    M,
    F,
}

export function isPlayerPosition(pos: string): pos is PlayerPositionType {
    return pos in playerPositionValues;
}

export enum PlayerQuality {
    'Terrible',
    'Very Bad',
    'Bad',
    'Low',
    'Passable',
    'Good',
    'Very Good',
    'Excellent',
    'Formidable',
    'World Class',
}

export type SkillsSummaries = {
    [position in PlayerPositionType]: SkillsSummary;
};

export interface PlayerWithSkillsSummaries extends Player {
    skillsSummaries?: SkillsSummaryCombo;
}

export interface SkillsSummaryWithBest extends SkillsSummaries {
    best?: SkillsSummary;
}

export interface FutureSkillsSummaryWithBest extends SkillsSummaryWithBest {
    options?: MatchesInSeasons;
}

export interface SkillsSummaryCombo {
    current?: SkillsSummaryWithBest;
    future?: FutureSkillsSummaryWithBest;
    futureMax?: FutureSkillsSummaryWithBest;
}

export interface Player {
    age: number;
    experience: PlayerQuality;
    experienceText: string;
    fitness: number;
    form?: string;
    id: number;
    maxTraining?: number;
    name: string;
    penalties: PlayerQuality;
    penaltiesText: string;
    position: string;
    positionZone: PlayerPositionType;
    potential: PlayerQuality;
    potentialText: string;
    quality: PlayerQuality;
    qualityText: string;
    skills: PlayerSkills;
    talent?: number;
    talentText?: string;
    temperament?: string | number; // TODO: enum
    value?: number;
    wage?: number;
}

export interface TlPlayer extends Player {
    deadline: string | number;
    price: string | number;
}
