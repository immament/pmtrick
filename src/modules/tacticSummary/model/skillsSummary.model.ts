import { PlayerPositionType } from '../../../common/model/player.model';

export interface SkillsSummary {
    main: number;
    secondary: number;
    tertiary: number;
    physical: number;
    gs: number;
    position: PlayerPositionType;
    rank?: number;
}

type SkillsGroups = 'main' | 'secondary' | 'physical' | 'tertiary';

export const skillsSummaryOrder: SkillsGroups[] = ['main', 'secondary', 'physical', 'tertiary'];
