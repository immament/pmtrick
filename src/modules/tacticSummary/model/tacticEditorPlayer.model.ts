import { StatsPositionKeys } from '@src/modules/tacticSummary/model/tacticStats.model';

import { Player } from '../../../common/model/player.model';

export interface TacticEditorPosition<T extends TacticEditorPlayer = TacticEditorPlayer> {
    position?: string;
    positionGroup?: StatsPositionKeys;
    id?: string;
    player?: T;
}

export interface TacticEditorData<T extends TacticEditorPlayer = TacticEditorPlayer> {
    positions: TacticEditorPosition<T>[];
}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface TacticEditorPlayer extends Player {}
