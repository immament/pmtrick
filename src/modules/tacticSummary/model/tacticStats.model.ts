import { PlayerPositionType } from '@src/common/model/player.model';
import { PlayerSkillsEnum, PlayerSkillsType } from '@src/common/model/playerSkills.model';

export enum StatsPosition {
    'all' = 'all',
    'g' = 'g',
    'd' = 'd',
    'm' = 'm',
    'a' = 'a',
}

const statsPositionKeyToPlayerPositionMap: Record<StatsPositionKeys, PlayerPositionType> = {
    g: 'G',
    d: 'D',
    m: 'M',
    a: 'F',
};

export function positionGroupToPlayerPosition(positionGroup: StatsPositionKeys): PlayerPositionType {
    return statsPositionKeyToPlayerPositionMap[positionGroup];
}

export type StatsPositionKeys = 'g' | 'd' | 'm' | 'a';
export const StatsPositionKeysArr = [...Object.keys(StatsPosition)] as StatsPosition[];

export type StatsKeys = PlayerSkillsType | 'gs';
export const StatsKeysArr = [...Object.keys(PlayerSkillsEnum), 'gs'] as StatsKeys[];

export type PositionStatRecord = { sum: number; avg?: number };

export type StatRecord = Record<StatsKeys, PositionStatRecord> & { count: number };

export type TacticStatsSums = Record<StatsPosition, StatRecord>;

export type AdvanceTacticTypes =
    | 'offside'
    | 'pressingLow'
    | 'pressingHigh'
    | 'counterAttack'
    | 'markingZonal'
    | 'markingMan'
    | 'highBalls'
    | 'oneOnOnes'
    | 'gkStand'
    | 'gkRush'
    | 'longShots'
    | 'firstTimeShots';
export type AdvanceTacticsValues = Record<AdvanceTacticTypes, (number | undefined)[]>;

export class TacticStats {
    ats!: AdvanceTacticsValues;
    vsAts!: AdvanceTacticsValues;

    constructor(statsSums: TacticStatsSums) {
        this.calculateAts(statsSums);
    }

    private calculateAts({ all, g, d, m, a }: TacticStatsSums): void {
        const dmCount = d.count + m.count;
        const dmTacklingAv = (d.tackling.sum + m.tackling.sum) / dmCount;
        const dmStrenghtAv = (d.strength.sum + m.strength.sum) / dmCount;
        const maCount = m.count + a.count;
        const maStrengthAv = (m.strength.sum + a.strength.sum) / maCount;

        const ats = {
            offside: [d.positioning.avg, d.speed.avg],
            pressingLow: [all.tackling.avg, all.speed.avg],
            pressingHigh: [all.passing.avg, all.speed.avg],
            counterAttack: [all.passing.avg, all.speed.avg],
            highBalls: [all.heading.avg, all.strength.avg],
            // Ats with one condition
            markingZonal: [dmTacklingAv, (d.speed.sum + m.speed.sum) / dmCount],
            markingMan: [dmTacklingAv, dmStrenghtAv],
            oneOnOnes: [(m.technique.sum + a.technique.sum) / maCount, maStrengthAv],
            gkStand: [g.reflexes.sum, g.handling.sum],
            gkRush: [g.agility.sum, g.outOfArea.sum],
            longShots: [(m.finishing.sum + a.finishing.sum) / maCount, (m.technique.sum + a.technique.sum) / maCount],
            firstTimeShots: [a.finishing.avg, a.heading.avg],
        };

        this.countAtWithOneConditionAverages(ats);
        this.ats = ats;

        const maPositioningAv = (m.positioning.sum + a.positioning.sum) / maCount;

        const vsAts = {
            offside: [a.positioning.avg, a.speed.avg],
            pressingLow: [all.tackling.avg, all.speed.avg],
            pressingHigh: [all.passing.avg, all.speed.avg],
            counterAttack: [all.passing.avg, all.speed.avg],
            highBalls: [all.heading.avg, all.strength.avg],
            // Ats with one condition
            markingZonal: [maPositioningAv, (m.speed.sum + a.speed.sum) / maCount],
            markingMan: [maPositioningAv, maStrengthAv],
            oneOnOnes: [dmTacklingAv, dmStrenghtAv],
            gkStand: [a.finishing.avg, a.heading.avg],
            gkRush: [a.technique.avg, a.heading.avg],
            longShots: [g.agility.avg, d.positioning.avg],
            firstTimeShots: [g.reflexes.avg, d.heading.avg],
        };

        this.countAtWithOneConditionAverages(vsAts);
        this.vsAts = vsAts;
    }

    private countAtWithOneConditionAverages(ats: AdvanceTacticsValues) {
        this.countAtAverage(ats.markingZonal);
        this.countAtAverage(ats.markingMan);
        this.countAtAverage(ats.oneOnOnes);
        this.countAtAverage(ats.gkStand);
        this.countAtAverage(ats.gkRush);
        this.countAtAverage(ats.longShots);
        this.countAtAverage(ats.firstTimeShots);
    }

    private countAtAverage(at: (number | undefined)[]): void {
        if (at[0] != undefined && at[1] != undefined) {
            at[2] = (at[0] + at[1]) / 2;
        } else {
            at[2] = 0;
        }
    }
}
