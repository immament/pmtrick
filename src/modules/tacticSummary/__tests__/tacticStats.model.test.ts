import { TacticStats } from '../model/tacticStats.model';

import { tacticSumsPropsMock } from './data/tacticSumsProps.data';

describe('Tactic Stats', () => {
    it('should calculate stats...', () => {
        const result = new TacticStats(tacticSumsPropsMock.getSums());
        expect(result).toMatchSnapshot();
    });
});
