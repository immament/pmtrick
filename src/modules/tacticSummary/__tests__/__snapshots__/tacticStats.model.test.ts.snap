// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tactic Stats should calculate stats... 1`] = `
TacticStats {
  "ats": Object {
    "counterAttack": Array [
      10.55,
      12.36,
    ],
    "firstTimeShots": Array [
      17.75,
      9.5,
      13.625,
    ],
    "gkRush": Array [
      13,
      20,
      16.5,
    ],
    "gkStand": Array [
      9,
      16,
      12.5,
    ],
    "highBalls": Array [
      7.45,
      8.91,
    ],
    "longShots": Array [
      10.857142857142858,
      9.857142857142858,
      10.357142857142858,
    ],
    "markingMan": Array [
      8.666666666666666,
      7.666666666666667,
      8.166666666666666,
    ],
    "markingZonal": Array [
      8.666666666666666,
      13.833333333333334,
      11.25,
    ],
    "offside": Array [
      14.67,
      9.67,
    ],
    "oneOnOnes": Array [
      9.857142857142858,
      8.714285714285714,
      9.285714285714285,
    ],
    "pressingHigh": Array [
      10.55,
      12.36,
    ],
    "pressingLow": Array [
      6.82,
      12.36,
    ],
  },
  "vsAts": Object {
    "counterAttack": Array [
      10.55,
      12.36,
    ],
    "firstTimeShots": Array [
      9,
      9.67,
      9.335,
    ],
    "gkRush": Array [
      12,
      9.5,
      10.75,
    ],
    "gkStand": Array [
      17.75,
      9.5,
      13.625,
    ],
    "highBalls": Array [
      7.45,
      8.91,
    ],
    "longShots": Array [
      13,
      14.67,
      13.835,
    ],
    "markingMan": Array [
      9.428571428571429,
      8.714285714285714,
      9.071428571428571,
    ],
    "markingZonal": Array [
      9.428571428571429,
      13.285714285714286,
      11.357142857142858,
    ],
    "offside": Array [
      8.5,
      9.75,
    ],
    "oneOnOnes": Array [
      8.666666666666666,
      7.666666666666667,
      8.166666666666666,
    ],
    "pressingHigh": Array [
      10.55,
      12.36,
    ],
    "pressingLow": Array [
      6.82,
      12.36,
    ],
  },
}
`;
